tests/core/pyspec/eth2spec/test/capella/sanity/test_blocks.py

```diff
@@ -255,3 +268,144 @@ def test_invalid_withdrawal_fail_second_block_payload_isnt_compatible(spec, stat
+    assert state.balances[validator_index] > 0
+    next_epoch_via_block(spec, state)
+    assert state.balances[validator_index] == 0
+    assert state.validators[validator_index].effective_balance > 0
+    next_epoch_via_block(spec, state)
+    assert state.validators[validator_index].effective_balance == 0
+
+    # Make a top-up balance to validator
+    amount = spec.MAX_EFFECTIVE_BALANCE // 4
+    deposit = prepare_state_and_deposit(spec, state, validator_index, amount, signed=True)
+
+    yield 'pre', state
+
+    block = build_empty_block_for_next_slot(spec, state)
+    block.body.deposits.append(deposit)
+
+    signed_block = state_transition_and_sign_block(spec, state, block)
+
+    yield 'blocks', [signed_block]
+    yield 'post', state
+
+    validator = state.validators[validator_index]
+    balance = state.balances[validator_index]
+    current_epoch = spec.get_current_epoch(state)
+    assert spec.is_fully_withdrawable_validator(validator, balance, current_epoch)
+
+
+def _insert_validator(spec, state, balance):
+    effective_balance = balance if balance < spec.MAX_EFFECTIVE_BALANCE else spec.MAX_EFFECTIVE_BALANCE
+    validator_index = len(state.validators)
+    validator = spec.Validator(
+        pubkey=pubkeys[validator_index],
+        withdrawal_credentials=spec.ETH1_ADDRESS_WITHDRAWAL_PREFIX + b'\x00' * 11 + b'\x56' * 20,
+        activation_eligibility_epoch=1,
+        activation_epoch=2,
+        exit_epoch=spec.FAR_FUTURE_EPOCH,
+        withdrawable_epoch=spec.FAR_FUTURE_EPOCH,
+        effective_balance=effective_balance,
+    )
+    state.validators.append(validator)
+    state.balances.append(balance)
+    state.previous_epoch_participation.append(spec.ParticipationFlags(0b0000_0000))
+    state.current_epoch_participation.append(spec.ParticipationFlags(0b0000_0000))
+    state.inactivity_scores.append(0)
+
+    return validator_index
+
+
+def _run_activate_and_partial_withdrawal(spec, state, initial_balance):
+    validator_index = _insert_validator(spec, state, balance=initial_balance)
+
+    next_epoch(spec, state)
+    transition_to(spec, state, spec.compute_start_slot_at_epoch(2) - 1)
+
+    assert not spec.is_active_validator(state.validators[validator_index], spec.get_current_epoch(state))
+
+    yield 'pre', state
+
+    blocks = []
+    # To activate
+    signed_block = transition_to_slot_via_block(spec, state, state.slot + 1)
+    blocks.append(signed_block)
+
+    assert spec.is_active_validator(state.validators[validator_index], spec.get_current_epoch(state))
+
+    if initial_balance > spec.MAX_EFFECTIVE_BALANCE:
+        assert spec.is_partially_withdrawable_validator(
+            state.validators[validator_index], state.balances[validator_index])
+    else:
+        assert not spec.is_partially_withdrawable_validator(
+            state.validators[validator_index], state.balances[validator_index])
+
+    # Getting attester rewards and getting partial withdrawals
+    for _ in range(2):
```

Comment:
What is the 2 magic number? to go through all validators and thus partially withdraw this validator?

Reply:
I was thinking of 1 epoch for balance updates, 1 epoch for withdrawals + balance updates.